File: services/web/src/lib/wiki.ts
import type { RelayFace } from "../api";

// The wiki (docs/design/21): the platform's shared, citable knowledge. A page
// is a row, a `[[slug]]` is how anything points at it, and a link to a page
// nobody has written yet is a feature — the wanted list is the wiki's to-do.
//
// The rules here are the browser's half of `agentplatform/wiki.py`. They have
// to agree digit for digit with it: the backend decides what counts as a link
// when it stores one, and this decides what the reader is looking at. A chip
// on a slug the backend never recorded is a page claiming a citation nobody
// made.

// The slug grammar, from the backend's SLUG_RE. Anchored whole: the slug goes
// into a URL and into `[[...]]`, so a partial match is not a slug.
export const SLUG_RE = /^[a-z0-9][a-z0-9-]{0,63}$/;
export const SLUG_MAX = 64;
// Untouched this long and the page is stale (the backend's `wiki_stale_days`
// default). Read client-side because the list route can only express "changed
// since", never "older than" — see `staleBefore` below.
export const STALE_DAYS = 30;

export type WikiPage = {
  id: string;
  slug: string;                 // the URL and the `[[link]]` target
  title: string;
  body: string;
  summary: string;              // the first paragraph, recomputed on every write
  tags: string[];
  version: number;
  created_by: string;
  updated_by: string;
  source_memory_id: string | null;   // set when the page was promoted
  created_at: string | null;
  updated_at: string | null;
  archived_at: string | null;
  // Attached by the API, never stored — the SSE payload carries the page
  // WITHOUT it, so anything drawing a face falls back to lib/face.
  updated_by_face?: RelayFace | null;
};

export type WikiPageRef = { slug: string; title: string };

export type WikiCitation = {
  message_id: string; channel_id: string; author: string; created_at: string | null;
};

export type WikiCitations = {
  count: number;
  // The scan behind `count` hit its limit, so the count is a floor: "200+".
  count_capped: boolean;
  last: WikiCitation[];
};

export type WikiPageDetail = {
  page: WikiPage;
  backlinks: WikiPageRef[];
  cited_in: WikiCitations;
};

export type WikiVersion = {
  id: string; page_id: string; version: number; title: string; body: string;
  author: string; run_id: string | null; reason: string; created_at: string | null;
  author_face?: RelayFace | null;
};

export type WikiHistoryRow = Omit<WikiVersion, "body" | "page_id"> & {
  added: number; removed: number;
};

export type WikiDiff = {
  version: WikiVersion; diff: string; added: number; removed: number;
};

export type WikiWantedRow = { slug: string; linked_from: string[] };

export type WikiAuthorCount = { author: string; count: number; face: RelayFace | null };

export type WikiStats = {
  pages: number;
  edits_24h: WikiAuthorCount[];
  wanted: number;
  stale: number;
  budget: { limit: number; agents: { agent: string; used: number; left: number }[] };
};

/** One line of `wiki.events`, which is what the SSE stream carries whole: the
 * page AND what the edit was, because "recent changes" is a list of edits and
 * the page alone says nothing about who changed it or why. */
export type WikiEvent = {
  event: string;                // created | edited | appended | archived | restored | promoted
  page: WikiPage;
  version: number;
  author: string;
  run_id: string | null;
  reason: string;
  added: number;
  removed: number;
};

export type WikiRef = { slug: string; start: number; end: number };

/** A fenced block blanked out character for character, so offsets still point
 * into the ORIGINAL text. Inline spans are handled by the chip pass, which
 * protects far more than fences (`lib/chips.ts`); this keeps `wikiRefs` honest
 * on its own for callers that hand it raw prose. */
function withoutFences(text: string): string {
  return text.replace(/```[\s\S]*?(?:```|$)/g, (block) => " ".repeat(block.length));
}

/** Every `[[slug]]` in a piece of text, as offsets into it.
 *
 * A target that is not a slug is not a link — it is words that happen to sit
 * in brackets, and chipping them would fill the wanted list with prose nobody
 * meant to write. This is `wiki.find_links`, minus the de-duplication: a body
 * with the same page named twice gets two chips. */
export function wikiRefs(text: string): WikiRef[] {
  if (!text) return [];
  const scanned = withoutFences(text);
  const out: WikiRef[] = [];
  for (const m of scanned.matchAll(/\[\[([^[\]\n]*)\]\]/g)) {
    const slug = m[1].toLowerCase();
    if (!SLUG_RE.test(slug)) continue;
    out.push({ slug, start: m.index, end: m.index + m[0].length });
  }
  return out;
}

/** The slug for a title — `Kyle's Location!` → `kyles-location`.
 *
 * The backend's `slugify`, which promotion and the "new page" form both have
 * to land on for the same words, or a page created here and a page promoted
 * there would be two pages about one thing. Accents fold (a slug is typed and
 * grepped, not read aloud); anything else becomes a single `-`. Empty when
 * nothing usable is left — the caller decides whether that is an error. */
export function slugify(title: string): string {
  const ascii = (title || "").normalize("NFKD")
    // Combining marks, then the apostrophes, so `Kyle's` is `kyles` and not
    // `kyle-s`. Both the typewriter one and the typographic one.
    .replace(/[\u0300-\u036f]/g, "").replace(/['’]/g, "").toLowerCase();
  return ascii.replace(/[^a-z0-9]+/g, "-").replace(/^-+|-+$/g, "")
    .slice(0, SLUG_MAX).replace(/-+$/, "");
}

/** A body's headings pushed down one level, outside fenced code.
 *
 * The page's own title is already the H1 on screen, and agents habitually open
 * a page with `# Its Title` — rendered as written, every page would have two
 * H1s and an outline that starts over halfway down. The markdown is untouched
 * in storage: this is how the document is SHOWN inside a page that already has
 * a heading, the same call the help pages make by having no title of their own.
 * `#` in a fenced block is a shell comment and stays one. */
export function demoteHeadings(body: string): string {
  return (body || "").replace(/```[\s\S]*?(?:```|$)|^(#{1,5})(?=\s)/gm,
                              (whole, hashes: string | undefined) => (hashes ? `#${hashes}` : whole));
}

/** The instant before which a page counts as stale.
 *
 * Client-side because the list route can only ask for pages changed SINCE a
 * time, never for ones that have not been touched since — so the stale list is
 * read off the (capped) listing rather than fetched. The count in the header
 * is the API's own, computed over every live page, which is why the two can
 * legitimately differ on a wiki bigger than the cap. */
export function staleBefore(now: number = Date.now()): number {
  return now - STALE_DAYS * 86400000;
}

export function isStale(page: WikiPage, now: number = Date.now()): boolean {
  const t = new Date(page.updated_at ?? "").getTime();
  return !Number.isNaN(t) && t < staleBefore(now);
}

/** A write that failed, with the parts a UI can act on. `detail` is the
 * sentence the API wrote; a 409 additionally carries where the page has got to,
 * which is what makes "reload and keep my text" possible. */
export class WikiWriteError extends Error {
  status: number;
  currentVersion: number | null;
  currentSummary: string;

  constructor(status: number, detail: string, current?: {
    version?: number | null; summary?: string;
  }) {
    super(detail);
    this.name = "WikiWriteError";
    this.status = status;
    this.currentVersion = current?.version ?? null;
    this.currentSummary = current?.summary ?? "";
  }
}

/** One write against the wiki.
 *
 * Deliberately not `api()`: that helper flattens a failure into
 * `"409: {json…}"`, and the 409 body is the whole point here — the version the
 * page is at now and what it says. A 401 still sends the reader to the login
 * page, the way every other call does. */
async function write<T>(path: string, method: string, body?: unknown): Promise<T> {
  const res = await fetch(path, {
    method, credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  if (res.status === 401) {
    window.location.href = "/login";
    throw new WikiWriteError(401, "not signed in");
  }
  const text = await res.text();
  let parsed: unknown = null;
  try { parsed = text ? JSON.parse(text) : null; } catch { /* not json */ }
  if (res.ok) return parsed as T;
  const payload = (parsed ?? {}) as Record<string, unknown>;
  // FastAPI answers a validation failure with a LIST under `detail`; a string
  // is what every refusal this UI can cause looks like. Anything else is
  // reported as the status, never as a JSON blob on screen.
  const detail = typeof payload.detail === "string"
    ? payload.detail : `the wiki refused this write (${res.status})`;
  throw new WikiWriteError(res.status, detail, {
    version: typeof payload.current_version === "number" ? payload.current_version : null,
    summary: typeof payload.current_summary === "string" ? payload.current_summary : "",
  });
}

/** What a human types into the editor. `tags` is sent as given — on a write
 * the API reads `null` as "leave them alone", which the editor never means:
 * what is in the field IS the page's tags. */
export type WikiDraft = { body: string; title: string; reason: string; tags: string[] };

export function createPage(slug: string, draft: WikiDraft): Promise<WikiPage> {
  return write<WikiPage>("/api/wiki/pages", "POST", { slug, ...draft });
}

export function savePage(slug: string, draft: WikiDraft, baseVersion: number):
  Promise<WikiPage> {
  return write<WikiPage>(`/api/wiki/pages/${encodeURIComponent(slug)}`, "PUT",
                         { ...draft, base_version: baseVersion });
}

export function restoreVersion(slug: string, version: number): Promise<WikiPage> {
  return write<WikiPage>(`/api/wiki/pages/${encodeURIComponent(slug)}/restore`, "POST",
                         { version, reason: `restored v${version}` });
}

export type WikiDiffLine = { text: string; kind: "add" | "del" | "hunk" | "meta" | "" };

/** A unified diff split into lines that can be coloured — and, more
 * importantly, PREFIXED: the `+` and the `-` are the diff's own characters, so
 * what an added line is never depends on being able to see the colour. */
export function diffLines(diff: string): WikiDiffLine[] {
  return (diff || "").split("\n").map((text) => ({ text, kind: lineKind(text) }));
}

function lineKind(line: string): WikiDiffLine["kind"] {
  if (line.startsWith("+++") || line.startsWith("---")) return "meta";
  if (line.startsWith("@@")) return "hunk";
  if (line.startsWith("+")) return "add";
  if (line.startsWith("-")) return "del";
  return "";
}

File: services/backend/agentplatform/wiki.py
"""The wiki (docs/design/21) is the platform's shared, citable knowledge: a page
is a row, a `[[slug]]` is how anything points at it, and every write is a diff
posted into `#wiki`. This module is its pure layer — the slug grammar, link
extraction, the summary, the diff, staleness, and every sentence the platform
writes into a room about a page — with no I/O.

It is separate from the store for the same reason `tickets.py` is: the API, the
`wiki` tool, the prompt block, the UI and the tests must agree on what a slug
is, what counts as a link and what the room was told, and a rule that lives in
one place cannot drift between them. The strings here are read by humans in the
web UI and, unrendered, in the Discord mirror, so they are golden-tested."""
import difflib
import re
import unicodedata
from datetime import datetime, timedelta, timezone

# Somebody else's words, flattened onto one line with room mentions defanged and
# a visible cap — the wiki quotes titles, reasons and authors into sentences it
# writes in its own voice exactly as Tickets does, and one sanitiser is one
# answer to what those words may contain.
from .tickets import one_line, participant_label
# Code is quoted text, never a link — a page explaining the `[[slug]]` syntax
# in a snippet (the seeded Home page does exactly that) must not create a wanted
# page for its example. Relay already decided what counts as code, fenced and
# inline, including the unclosed fence an LLM truncated mid-block; borrowing its
# patterns keeps "what is code" one answer across the platform.
from .relay import _CODE_RE, _FENCE

# The slug grammar (docs/design/21). Anchored with \A..\Z rather than ^..$: the
# slug is interpolated into URLs and into `[[...]]`, and `$` would let
# "home\nrm -rf" through a caller that reached for `.match`. 64 characters is
# the column width.
SLUG_RE = re.compile(r"\A[a-z0-9][a-z0-9-]{0,63}\Z")
SLUG_MAX = 64
# Caps matching the columns these fields are stored in, and the point past
# which none of them is information any more: a title is a headline, a reason
# is a sentence, a summary is the first paragraph.
TITLE_LIMIT, REASON_LIMIT, SUMMARY_LIMIT, LABEL_LIMIT = 120, 200, 280, 64
# Same shape as `relay_router.BUDGET_PREFIX` and Tickets': the store finds "did
# we already say this in this room this hour?" with a LIKE on the prefix, so the
# variable part has to come last.
BUDGET_PREFIX = "⏸️ paused: an agent has used its hourly wiki-write budget"

_FENCE_RE = re.compile(_FENCE, re.DOTALL)
# A link target is whatever is between the brackets — it is validated against
# the grammar afterwards rather than matched loosely here, so `[[not a slug]]`
# is text and not a link to `not-a-slug` nobody typed.
_LINK_RE = re.compile(r"\[\[([^\[\]\n]*)\]\]")
# `[text](url)` unwrapped to `text` for a summary. The url is allowed one level
# of nested parentheses, which both a wikipedia link and `javascript:alert(1)`
# have — stopping at the first `)` would leave the tail of the url in the
# summary, which is the one string that is never rendered as markdown.
_MD_LINK_RE = re.compile(r"\[([^\[\]]*)\]\((?:[^()]|\([^()]*\))*\)")
# Heading markers at the start of any line of the paragraph a summary is taken
# from: `## Wanted` summarises as `Wanted`.
_HEADING_RE = re.compile(r"^[ \t]*#{1,6}[ \t]*", re.MULTILINE)
_PARA_RE = re.compile(r"\n[ \t]*\n")
# Apostrophes vanish rather than becoming separators, so `Kyle's` slugs as
# `kyles` and not `kyle-s`. Both the typewriter and the typographic one: the
# ASCII fold below would drop the curly one anyway, but only by accident.
_APOSTROPHE_RE = re.compile(r"['’ʼ]")
_NON_SLUG_RE = re.compile(r"[^a-z0-9]+")


def _text(body) -> str:
    r"""A body as the parsers see it. CRLF is normalised because a paragraph
    break is a blank line and `\r\n\r\n` is not one to any regex that says
    `\n`: a body pasted from Windows or fetched over HTTP would be one
    unsplittable paragraph, and its summary the whole page."""
    return (body or "").replace("\r\n", "\n").replace("\r", "\n")


def _is_heading_only(para: str) -> bool:
    """A paragraph that is nothing but headings — the shape a summary skips
    past. A heading with prose under it in the same paragraph is prose."""
    lines = [line for line in para.splitlines() if line.strip()]
    return bool(lines) and all(line.lstrip().startswith("#") for line in lines)


def slugify(title_or_key) -> str:
    """The slug for a title or a memory key: `Kyle's Location!` → `kyles-location`.

    Deterministic and lossy on purpose — promotion derives a slug from a memory
    key and a human creating a page gets one offered from the title, and both
    have to land on the same answer for the same words. Accents fold to ASCII
    (a slug is typed and grepped, not read aloud) and anything that is not a
    letter or a digit becomes a single `-`. Raises when nothing usable is left:
    a page identified by the empty string is a page nobody can link to."""
    folded = unicodedata.normalize("NFKD", str(title_or_key or ""))
    ascii_only = folded.encode("ascii", "ignore").decode("ascii").lower()
    slug = _NON_SLUG_RE.sub("-", _APOSTROPHE_RE.sub("", ascii_only)).strip("-")
    # Trimmed again after the cap: cutting mid-word can land on a separator, and
    # a trailing `-` is not a slug the grammar accepts.
    slug = slug[:SLUG_MAX].strip("-")
    if not SLUG_RE.fullmatch(slug):
        raise ValueError(f"no usable wiki slug in {title_or_key!r}")
    return slug


def find_links(body: str) -> list[str]:
    """The `[[slug]]` targets a body points at: first-seen order, lower-cased,
    no duplicates. A target that is not a slug is not a link — it is text that
    happens to sit in brackets, and turning it into a wanted page would fill the
    red-link list with prose nobody meant to write.

    A target that does not exist yet IS a link: that dangling row is the wanted
    page, which is the feature.

    Inline spans count as code here, unlike `tickets.find_ticket_refs` where
    naming a ticket in backticks is still naming it: a `[[slug]]` in backticks
    is someone showing the syntax, and the web renderer's chip pass already
    treats an inline span as inert, so linking it would disagree with what the
    reader is looking at."""
    out: list[str] = []
    for m in _LINK_RE.finditer(_CODE_RE.sub(" ", _text(body))):
        target = m.group(1).lower()
        if SLUG_RE.fullmatch(target) and target not in out:
            out.append(target)
    return out


def summary_of(body: str) -> str:
    """The page's first paragraph, on one line — what search results, the
    citation chip and every agent's `<wiki>` prompt block show instead of a
    64 KB body.

    A page that opens with its own title as a heading would summarise as that
    title, which says nothing the slug did not, so the heading is skipped when
    there is a paragraph behind it. Markdown that only means something rendered
    is unwrapped (`[[deploying]]` → `deploying`, `[the docs](url)` → `the docs`)
    because this string is shown as plain text everywhere it appears — and it
    goes through the same flattening as any other quoted words: the body is
    agent-written, and the summary is replayed into other agents' prompts."""
    # Fenced code is dropped before anything else: a page that opens with a
    # snippet would otherwise summarise as ``` plus a shell command, in search
    # results and in every other agent's prompt block. Inline spans stay — they
    # read as ordinary words in a one-line summary.
    paras = [p for p in _PARA_RE.split(_FENCE_RE.sub(" ", _text(body))) if p.strip()]
    if not paras:
        return ""
    # Every leading heading, not just the first: `# Page` over `## Section` over
    # the prose is a common shape, and stopping at the first hop summarises the
    # page as the word "Section".
    i = 0
    while i < len(paras) - 1 and _is_heading_only(paras[i]):
        i += 1
    text = _HEADING_RE.sub("", paras[i])
    text = _LINK_RE.sub(lambda m: m.group(1), text)
    text = _MD_LINK_RE.sub(lambda m: m.group(1), text)
    return one_line(text, SUMMARY_LIMIT)


def unified_diff(old: str, new: str, slug: str, v_old, v_new) -> str:
    """The text diff between two versions of a page, headed with the versions it
    is between. Versions store the full body precisely so this can be computed
    on read: nothing about a page's history depends on a patch having been
    stored correctly at write time."""
    return "\n".join(difflib.unified_diff(
        (old or "").splitlines(), (new or "").splitlines(),
        fromfile=f"{slug} v{v_old}", tofile=f"{slug} v{v_new}", lineterm=""))


def line_counts(old: str, new: str) -> tuple[int, int]:
    """`(added, removed)` for one write — the `(+12 −1)` on the diff card and in
    the history list. The two header lines are dropped by position rather than
    by prefix: a page that quotes a diff has body lines starting with `+++`, and
    they are content."""
    diff = list(difflib.unified_diff((old or "").splitlines(),
                                     (new or "").splitlines(), lineterm=""))
    added = sum(1 for line in diff[2:] if line.startswith("+"))
    removed = sum(1 for line in diff[2:] if line.startswith("-"))
    return added, removed


def card_for(page, version_row, *, url_base: str, added: int = 0,
             removed: int = 0) -> dict:
    """The `card` payload on a write's event message in `#wiki`. Kept small and
    flat: the Relay client renders it directly, and the Discord bridge falls
    back to the body below when it cannot. The line counts are passed in
    because they are a fact about two bodies and the card only has one."""
    return {"type": "wiki", "slug": page.slug,
            "title": one_line(page.title, TITLE_LIMIT),
            "version": version_row.version,
            # Kept namespaced (`agent:pai`) so the client can resolve a face,
            # but flattened all the same: a connector id is not the platform's
            # own text either.
            "author": one_line(version_row.author, LABEL_LIMIT),
            "reason": one_line(version_row.reason, REASON_LIMIT),
            "added": added, "removed": removed,
            "url": f"{(url_base or '').rstrip('/')}/wiki/{page.slug}"}


def card_body(page, version_row, *, added: int = 0, removed: int = 0) -> str:
    """The card message's plain text. It exists because the Discord mirror (and
    any other bridge) gets the body and not the card, so the sentence has to
    carry the edit on its own — and it keeps the `[[slug]]` so the mirror reads
    as a link back into the wiki. An empty reason drops the quotes rather than
    printing a pair with nothing between them."""
    who = one_line(participant_label(version_row.author), LABEL_LIMIT)
    reason = one_line(version_row.reason, REASON_LIMIT)
    said = f'{who}: "{reason}"' if reason else who
    return (f"📖 [[{page.slug}]] v{version_row.version} · {said} "
            f"(+{added} −{removed})")


def budget_body(limit: int) -> str:
    return f"{BUDGET_PREFIX} ({limit}/hour); try again later"


def promoted_body(memory_content: str, agent: str, key, when) -> str:
    """A promoted memory's page body: the note, then where it came from. The
    provenance line is the whole point of promotion — the memory stays the
    agent's private note and the page becomes the shared fact, so the page has
    to say whose note it was and when it hardened. An unnamed memory (promotion
    by id) simply drops the key rather than quoting an empty one."""
    named = f" `{one_line(key, LABEL_LIMIT)}`" if key else ""
    line = (f"Promoted from {one_line(participant_label(agent), LABEL_LIMIT)}'s"
            f" memory{named} on {when:%Y-%m-%d}")
    content = (memory_content or "").rstrip()
    return f"{content}\n\n{line}" if content else line


def _aware(ts: datetime) -> datetime:
    return ts.replace(tzinfo=timezone.utc) if ts.tzinfo is None else ts


def is_stale(page, now: datetime, days: int) -> bool:
    """A live page nobody has touched for `days` — the gardener's list and the
    stats. Archived pages are out of the garden by definition: archiving is how
    a page stops being something anyone has to tend."""
    if page.archived_at is not None or page.updated_at is None:
        return False
    # sqlite hands back naive datetimes and postgres does not, and a caller
    # reaching for `utcnow()` passes a naive `now`. Either mismatch is a
    # TypeError that would take down a whole gardener pass over one row, so both
    # operands are normalised rather than one.
    return (_aware(now) - _aware(page.updated_at)) >= timedelta(days=days)
